network/src/compress.rs

```diff
@@ -0,0 +1,160 @@
+use bytes::{Bytes, BytesMut};
+use log::debug;
+use snap::{Decoder as SnapDecoder, Encoder as SnapEncoder};
+use tokio::codec::{Decoder, Encoder, LengthDelimitedCodec};
+
+use std::io;
+
+const SKIP_COMPRESS_SIZE: usize = 40 * 1024;
+
+/// Compressed decompression structure
+///
+/// If you want to support multiple compression formats in the future,
+/// you can simply think that 0b1000 is in snappy format and 0b0000 is in uncompressed format.
+///
+/// # Message in Bytes:
+///
+/// +---------------------------------------------------------------+
+/// | Bytes | Type | Function                                       |
+/// |-------+------+------------------------------------------------|
+/// |   0   |  u1  | Compress: true 1, false 0                      |
+/// |       |  u7  | Reserved                                       |
+/// +-------+------+------------------------------------------------+
+/// |  1~   |      | Payload (Serialized Data with Compress)        |
+/// +-------+------+------------------------------------------------+
+#[derive(Clone, Debug)]
+struct Message {
+    inner: BytesMut,
+}
+
+impl Message {
+    /// create from raw data
+    fn from_raw(data: Bytes) -> Self {
+        let mut inner = BytesMut::from(&[0u8][..]);
+        inner.unsplit(BytesMut::from(data));
+        Self { inner }
+    }
+
+    /// create from compressed data
+    fn from_compressed(data: BytesMut) -> Self {
+        Self { inner: data }
+    }
+
+    /// Compress message
+    fn compress(mut self) -> Bytes {
+        let input = self.inner.split_off(1);
```

Comment:
Checking the bytes len first, which avoids `split/unsplit` for small size message.
```
if self.inner.len() > SKIP_COMPRESS_SIZE {
}
```